### PR TITLE
Music Cards

### PR DESCRIPTION
**Pull request by McPolemic**
This is a quick cross-post from my horrifically-neglected personal blog. I'm not super-jazzed by how small the Vimeo embed is, but embedding via YouTube resulted in a full-width, 60px high video.

¯\\\_(ツ)\_/¯

### CODE REVIEW

File: app/posts/2018-02-20-music-cards.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+tldr:
+  body: """
+        Software doesn't always have to be amazing. Sometimes, it can just be useful.
+        """
+reddit: false
+---
+
+# Music Cards
+
+Our family loves music. We almost have some sort of music going on in the background at our house. We have a number of Sonos speakers and subscriptions to the major music providers, giving us a lot of flexibility in what to listen to and where.
```
Comment (marcpeabody) [inline app/posts/2018-02-20-music-cards.md]: almost always?

File: app/posts/2018-02-20-music-cards.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+This file has three classes:
+* Spotify
+* Jukebox 
+* CodeReader
+
+`Spotify` is a wrapper for the `rspotify` gem. It handles searching for individual songs and playlists. Searching takes between 1/10th and 1/4 of a second, which can add up, so it also handles caching the results locally. Once we have the result saved, it takes around 1/1000th of a second.
+
+`Jukebox` wraps all interactions with the Sonos speakers. It takes in a speaker name and a Spotify wrapper. Notably, because I ran into some problems getting the `sonos` gem to work with Spotify URLs, it handles the less-than-pleasant encoding workarounds.
+
+Finally, `CodeReader` orchestrates the whole thing. It has a dictionary that maps between our unique codes (e.g. '08931021') and the action we'd like the jukebox to play (`jukebox.play_spotify_track('Remember Me')`). By using Spotify to search for the track instead of encode specific tracks, I've (hopefully) made it easier to read and easier to manage down the road as we add more songs and albums. It runs forever in the `repl` method, which waits for input, runs the resulting action, and repeats.
```
Comment (Schoonology) [inline app/posts/2018-02-20-music-cards.md]: Good song choice! (And what an _amazing_ film!)

Comment (Schoonology) [reply app/posts/2018-02-20-music-cards.md]: instead of _encoding_ specific tracks

File: app/posts/2018-02-20-music-cards.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+* I didn't realize RFID cards and readers have different frequencies and bought the wrong reader at first. I returned it and bought the right one.
+* I thought I could always read from STDIN. Only after assembling it did I realize I'd have to read directly from the keyboard.
+* Sonos apparently rolled out some encrypted method of retrieving metadata from Spotify, breaking third-party integration. I don't really understand this, but it means that I have to escape the URL and do without song/artist name in the Sonos controller. In practice, this isn't a big deal.
+* I was really worried about the responsiveness of this. If Nika scanned a card and it played two seconds later, it would feel broken. I ended up adding a lot of timing logging and caching to ensure it feels snappy.
+* I originally used Nika's glitter glue to glue the printed out pictures to the RFID cards. You can imagine my surprise when I realized glitter glue doesn't work super well. Luckily, Liz has more craft experience than I do and had some recommendations.
```
Comment (Schoonology) [inline app/posts/2018-02-20-music-cards.md]: 😂

File: app/posts/2018-02-20-music-cards.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+I used Docker so I could develop on my fast desktop and then deploy it to the Raspberry Pi. I didn't realize/remember that they are different platforms (x86 vs ARM), which comes with a whole mess of problems. I still like Docker for packaging, but I should've moved it sooner onto the Raspberry Pi.
```
Comment (Schoonology) [inline app/posts/2018-02-20-music-cards.md]: Missing bullet.

File: app/posts/2018-02-20-music-cards.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+The full code is located here: [https://github.com/McPolemic/nika_tunes/](https://github.com/McPolemic/nika_tunes/)
+
+It consists of two parts: a [`reader.py`](https://github.com/McPolemic/nika_tunes/blob/master/reader.py) that connects to the USB reader and outputs the serial numbers, and [`nika_tunes.rb`](https://github.com/McPolemic/nika_tunes/blob/master/nika_tunes.rb), which reads in the unique codes, looks up songs via Spotify, and plays them.
+
+I didn't realize I'd need `reader.py` at first. I was going to read the codes from STDIN. Unfortunately, I didn't think ahead. When I started putting things together, I realized that I wanted this to work without logging in (which means no STDIN). I needed to read raw keys from the RFID reader, which means digging down into Linux's `evdev` system. I looked through Rubygems but didn't find a library that worked to my liking. Luckily, Python has quite good `evdev` support, so I wrote this small script to handle decoding keyboard events into strings.
+
+The output is piped into `nika_tunes.rb`, which has the bulk of the logic.
```
Comment (Schoonology) [inline app/posts/2018-02-20-music-cards.md]: If I want anything, it's more content. This is a _wonderful_ write-up. One place I'd want more discussion is here: the strength of using two languages/runtimes together. What a wise choice.

File: app/posts/2018-02-20-music-cards.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+This is Nika. She also loves music. Unfortunately, she does not yet know how to read, which is a real pain when she wants to pick out music. One option would be to buy CDs, but then she has to keep them from getting scratched and remember which tracks are the songs she likes. Also, we'd have to buy a huge amount of CDs (or burn a similar number of CDs) and a CD player when we already have three subscription music services and speakers throughout the house. I hoped for a better way.
+
+## Hardware
+
+What I wanted was something sturdier and cheaper than CDs. I don't really need them to _store_ audio. We have Spotify for that. I'd rather they just _identify_ what audio Nika would like to play.
+
+RFID cards were a good candidate than this. They're sturdy, relatively cheap (I bought 50 for $13), and have good support for USB readers. I bought [this RFID reader](https://www.amazon.com/gp/product/B018C8C162/ref=oh_aui_detailpage_o02_s00?ie=UTF8&psc=1) and [these RFID cards](https://www.amazon.com/gp/product/B00GXV4IGC/ref=oh_aui_detailpage_o03_s01?ie=UTF8&psc=1). I hooked them both up to a Raspberry Pi 3 running Raspbian.
```
Comment (Schoonology) [inline app/posts/2018-02-20-music-cards.md]: The only other place where I'd want more detail is here: what were your _alternatives_?

**Comment by Schoonology** [comment]
Please don't take my comments as "do this", rather as "if you were to keep editing this before publish, here's where I would edit".

**Comment by McPolemic** [comment]
@Schoonology Thanks for your great feedback 💚. I filled out a little bit based on your comments but I worry I've spilled enough ink as is on this 😄. I like leaving a little room for discussion if anyone ends up interested in the particulars.

File: app/posts/2018-02-20-music-cards.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+<iframe src="https://player.vimeo.com/video/256652780" width="640" height="360" frameborder="0" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe>
+
+We printed out pictures and glued them to the various RFID cards. I started out with nine to see how she felt.
+
+![The RFID reader and RFID cards with graphics glued on the front](https://dl.dropboxusercontent.com/s/gt4x6ntorsvbx1v/assembled%20reader.jpg?raw=1)
+![A text message from Liz stating, "Nika is really excited about her 'music cards'. :)"](https://dl.dropboxusercontent.com/s/xx9czjvjii7wq5h/nika_excited.jpg?raw=1)
+
+## Challenges
+
+Often, I feel like posts like this don't delve enough into the challenges someone can run into, making it feel like everything went swimmingly. I ran quite a few challenges, which are worth mentioning for a second.
```
Comment (rosston) [inline app/posts/2018-02-20-music-cards.md]: I believe you're missing a word here. Probably supposed to be "ran into" rather than just "ran"?

**Pull request merged**